// @shapediver/viewer.settings/dist/versions/v1/Validator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validate = void 0;
const joi_1 = __importDefault(require("joi"));
const lightSchema = joi_1.default.object({
    id: joi_1.default.string().optional(),
    name: joi_1.default.string().optional(),
    type: joi_1.default.string().optional(),
    properties: {
        color: joi_1.default.alternatives([joi_1.default.number().unsafe(), joi_1.default.string()]).optional(),
        direction: joi_1.default.object({ x: joi_1.default.number().unsafe(), y: joi_1.default.number().unsafe(), z: joi_1.default.number().unsafe() }).optional(),
        position: joi_1.default.object({ x: joi_1.default.number().unsafe(), y: joi_1.default.number().unsafe(), z: joi_1.default.number().unsafe() }).optional(),
        target: joi_1.default.object({ x: joi_1.default.number().unsafe(), y: joi_1.default.number().unsafe(), z: joi_1.default.number().unsafe() }).optional(),
        castShadow: joi_1.default.boolean().optional(),
        skyColor: joi_1.default.alternatives([joi_1.default.number().unsafe(), joi_1.default.string()]).optional(),
        groundColor: joi_1.default.alternatives([joi_1.default.number().unsafe(), joi_1.default.string()]).optional(),
        intensity: joi_1.default.number().unsafe().optional(),
        distance: joi_1.default.number().unsafe().optional(),
        angle: joi_1.default.number().unsafe().optional(),
        penumbra: joi_1.default.number().unsafe().optional(),
        decay: joi_1.default.number().unsafe().optional(),
        shadowMapResolution: joi_1.default.number().unsafe().optional(),
        shadowMapBias: joi_1.default.number().unsafe().optional()
    }
}).optional();
const lightScenesSchema = joi_1.default.object().pattern(/^/, [
    joi_1.default.object({
        id: joi_1.default.string(),
        name: joi_1.default.string().optional(),
        lights: joi_1.default.object().pattern(/^/, [lightSchema])
    })
]);
const schema = joi_1.default.object({
    build_date: joi_1.default.string().allow('').optional(),
    build_version: joi_1.default.string().allow('').optional(),
    settings_version: joi_1.default.string().optional(),
    ambientOcclusion: joi_1.default.boolean().optional(),
    autoRotateSpeed: joi_1.default.number().unsafe().min(0).optional(),
    backgroundColor: joi_1.default.string().optional(),
    bumpAmplitude: joi_1.default.number().unsafe().min(0).optional(),
    camera: joi_1.default.object({ position: joi_1.default.object({ x: joi_1.default.number().unsafe().optional(), y: joi_1.default.number().unsafe().optional(), z: joi_1.default.number().unsafe().optional() }).optional(), target: joi_1.default.object({ x: joi_1.default.number().unsafe().optional(), y: joi_1.default.number().unsafe().optional(), z: joi_1.default.number().unsafe().optional() }).optional() }).optional(),
    cameraAutoAdjust: joi_1.default.boolean().optional(),
    cameraMovementDuration: joi_1.default.number().unsafe().min(0).optional(),
    cameraOrtho: joi_1.default.object({ position: joi_1.default.object({ x: joi_1.default.number().unsafe().optional(), y: joi_1.default.number().unsafe().optional(), z: joi_1.default.number().unsafe().optional() }).optional(), target: joi_1.default.object({ x: joi_1.default.number().unsafe().optional(), y: joi_1.default.number().unsafe().optional(), z: joi_1.default.number().unsafe().optional() }).optional() }).optional(),
    cameraRevertAtMouseUp: joi_1.default.boolean().optional(),
    clearAlpha: joi_1.default.number().unsafe().min(0).max(1).optional(),
    clearColor: joi_1.default.string().optional(),
    commitParameters: joi_1.default.boolean().optional(),
    controlDamping: joi_1.default.number().unsafe().min(0).optional(),
    controlNames: joi_1.default.object().pattern(/.*/, [joi_1.default.string()]).optional(),
    controlOrder: joi_1.default.array().items(joi_1.default.string()).optional(),
    defaultMaterialColor: joi_1.default.alternatives([joi_1.default.string(), joi_1.default.array().items(joi_1.default.number().unsafe())]).optional(),
    disablePan: joi_1.default.boolean().optional(),
    disableZoom: joi_1.default.boolean().optional(),
    enableAutoRotation: joi_1.default.boolean().optional(),
    enableRotation: joi_1.default.boolean().optional(),
    environmentMap: joi_1.default.alternatives([joi_1.default.string(), joi_1.default.array().items(joi_1.default.string())]).optional(),
    environmentMapResolution: joi_1.default.string().optional(),
    fov: joi_1.default.number().unsafe().positive().max(180).optional(),
    lightScene: joi_1.default.string().optional(),
    lightScenes: lightScenesSchema.allow(null).optional(),
    panSpeed: joi_1.default.number().unsafe().min(0).optional(),
    parametersHidden: joi_1.default.array().items(joi_1.default.string()).optional(),
    pointSize: joi_1.default.number().unsafe().min(0).optional(),
    revertAtMouseUpDuration: joi_1.default.number().unsafe().min(0).optional(),
    rotateSpeed: joi_1.default.number().unsafe().min(0).optional(),
    showEnvironmentMap: joi_1.default.boolean().optional(),
    showGrid: joi_1.default.boolean().optional(),
    showGroundPlane: joi_1.default.boolean().optional(),
    showShadows: joi_1.default.boolean().optional(),
    topView: joi_1.default.boolean().optional(),
    zoomExtentFactor: joi_1.default.number().unsafe().min(0).optional(),
    zoomSpeed: joi_1.default.number().unsafe().min(0).optional(),
}).options({ allowUnknown: true });
const validate = (s) => {
    const result = joi_1.default.attempt(s, schema);
    s = result.value;
};
exports.validate = validate;
