// @shapediver/sdk.geometry-api-sdk-v2/dist/resources/ShapeDiverAssetApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverAssetApi = void 0;
const sdk_geometry_api_sdk_core_1 = require("@shapediver/sdk.geometry-api-sdk-core");
const utils_1 = require("../utils/utils");
const apiAssetExportUri = /.+\/session\/.+\/export\/.+/;
const apiAssetOutputUri = /.+\/session\/.+\/output\/.+/;
const apiAssetTextureUri = /.+\/session\/.+\/texture\/.+/;
const cdnAssetUri = /.+\/cdn-asset-(exports|outputs|textures)\/.+/;
const cdnAssetExportUri = /.+\/cdn-asset-exports\/.+/;
const cdnAssetOutputUri = /.+\/cdn-asset-outputs\/.+/;
const cdnAssetTextureUri = /.+\/cdn-asset-textures\/.+/;
class ShapeDiverAssetApi extends sdk_geometry_api_sdk_core_1.BaseResourceApi {
    constructor(api) {
        super(api);
    }
    /**
     * Download an export.
     *
     * @param sessionId
     * @param assetData
     */
    getExport(sessionId, assetData) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () {
                var _a;
                const [header, data] = yield this.api.get(`${this.buildSessionUri(sessionId)}/export/${assetData}`, { responseType: sdk_geometry_api_sdk_core_1.ShapeDiverSdkApiResponseType.DATA });
                const contentType = (_a = header["Content-Type"]) !== null && _a !== void 0 ? _a : header["content-type"];
                return [data, contentType];
            }));
        });
    }
    /**
     * Download an output.
     *
     * @param sessionId
     * @param assetData
     */
    getOutput(sessionId, assetData) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () {
                var _a;
                const [header, data] = yield this.api.get(`${this.buildSessionUri(sessionId)}/output/${assetData}`, { responseType: sdk_geometry_api_sdk_core_1.ShapeDiverSdkApiResponseType.DATA });
                const contentType = (_a = header["Content-Type"]) !== null && _a !== void 0 ? _a : header["content-type"];
                return [data, contentType];
            }));
        });
    }
    /**
     * Downloads the JSON content part of a sdTF output.
     *
     * @param sessionId
     * @param assetData
     */
    getSdtfJsonContent(sessionId, assetData) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () {
                return (yield this.api.get(`${this.buildSessionUri(sessionId)}/output/${assetData}`, {
                    responseType: sdk_geometry_api_sdk_core_1.ShapeDiverSdkApiResponseType.JSON,
                    accept: "model/vnd.sdtf+json",
                }))[1];
            }));
        });
    }
    /**
     * Download a texture file.
     *
     * @param sessionId
     * @param assetData
     */
    getTexture(sessionId, assetData) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () {
                var _a;
                const [header, data] = yield this.api.get(`${this.buildSessionUri(sessionId)}/texture/${assetData}`, { responseType: sdk_geometry_api_sdk_core_1.ShapeDiverSdkApiResponseType.DATA });
                const contentType = (_a = header["Content-Type"]) !== null && _a !== void 0 ? _a : header["content-type"];
                return [data, contentType];
            }));
        });
    }
    /**
     * Download a glTF file.
     *
     * @param sessionId
     * @param assetData
     */
    getGltf(sessionId, assetData) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () {
                return (yield this.api.get(`${this.buildSessionUri(sessionId)}/gltf/${assetData}`, { responseType: sdk_geometry_api_sdk_core_1.ShapeDiverSdkApiResponseType.DATA }))[1];
            }));
        });
    }
    /**
     * Download a USDZ file.
     *
     * @param sessionId
     * @param assetData
     */
    getUsdz(sessionId, assetData) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () {
                return (yield this.api.get(`${this.buildSessionUri(sessionId)}/usdz/${assetData}`, { responseType: sdk_geometry_api_sdk_core_1.ShapeDiverSdkApiResponseType.DATA }))[1];
            }));
        });
    }
    /**
     * Download an image.
     *
     * @param sessionId
     * @param url - The URL of the image that should be downloaded.
     * @returns Array of size 2: [0] = content data, [1] = content type.
     */
    downloadImage(sessionId, url) {
        return __awaiter(this, void 0, void 0, function* () {
            let targetUrl;
            if (apiAssetTextureUri.test(url) || cdnAssetTextureUri.test(url)) {
                // Call ShapeDiver texture-asset URLs directly
                targetUrl = url;
            }
            else {
                // All other source URLs are called via the download-image endpoint
                targetUrl = `${this.buildSessionUri(sessionId)}/image?url=${(0, utils_1.encodeBase64)(url)}`;
            }
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () {
                var _a;
                const [header, data] = yield this.api.get(targetUrl, {
                    responseType: sdk_geometry_api_sdk_core_1.ShapeDiverSdkApiResponseType.DATA,
                    disableAuthorization: cdnAssetUri.test(url), // disable for CDN URLs
                });
                const contentType = (_a = header["Content-Type"]) !== null && _a !== void 0 ? _a : header["content-type"];
                return [data, contentType];
            }));
        });
    }
    /**
     * Fetches a ShapeDiver asset of the following types:
     *  * Output
     *  * Export
     *  * Texture
     *
     * This function works similar to {@link getOutput}, {@link getExport} and {@link getTexture}, but does not require
     * extracted _session ID_ and _asset data_ parameters.
     *
     * @param url - The URL of the asset that should be fetched.
     * @returns Array of size 3: [0] = content data, [1] = content type, [2] = asset type.
     * @throws {@link ShapeDiverError} when the given URL is not a valid ShapeDiver asset URL.
     */
    getAsset(url) {
        return __awaiter(this, void 0, void 0, function* () {
            let type;
            // Check if the given URL is a valid API or CDN asset URL
            if (apiAssetExportUri.test(url) || cdnAssetExportUri.test(url))
                type = "export";
            else if (apiAssetOutputUri.test(url) || cdnAssetOutputUri.test(url))
                type = "output";
            else if (apiAssetTextureUri.test(url) || cdnAssetTextureUri.test(url))
                type = "texture";
            else
                throw new sdk_geometry_api_sdk_core_1.ShapeDiverError("Cannot fetch asset: Invalid URL (only ShapeDiver asset URLs are allowed).");
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () {
                var _a;
                const [header, data] = yield this.api.get(url, {
                    responseType: sdk_geometry_api_sdk_core_1.ShapeDiverSdkApiResponseType.DATA,
                    disableAuthorization: cdnAssetUri.test(url), // disable for CDN URLs
                });
                const contentType = (_a = header["Content-Type"]) !== null && _a !== void 0 ? _a : header["content-type"];
                return [data, contentType, type];
            }));
        });
    }
}
exports.ShapeDiverAssetApi = ShapeDiverAssetApi;
