// @shapediver/sdk.geometry-api-sdk-v2/dist/resources/ShapeDiverSystemApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverSystemApi = void 0;
const sdk_geometry_api_sdk_core_1 = require("@shapediver/sdk.geometry-api-sdk-core");
const utils_1 = require("../utils/utils");
class ShapeDiverSystemApi extends sdk_geometry_api_sdk_core_1.BaseResourceApi {
    constructor(api) {
        super(api);
    }
    /**
     * Log the specified message.
     *
     * @param sessionId
     * @param body
     */
    log(sessionId, body) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () { return (yield this.api.post(this.buildSessionUri(sessionId) + "/log/message", body))[1]; }));
        });
    }
    /**
     * Remove a model with all it's components.
     *
     * **WARNING: This operation cannot be undone!**
     *
     * @param modelId
     */
    pruneModel(modelId) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () { return (yield this.api.delete(`${this.buildSystemUri()}/model/${modelId}`))[1]; }));
        });
    }
    /**
     * Decrypt the given ticket.
     *
     * @param ticket
     */
    decryptTicket(ticket) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () { return (yield this.api.get(this.buildTicketUri(ticket)))[1]; }));
        });
    }
    /**
     * Create a new authentication group.
     *
     * @param body
     */
    authGroup(body) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () { return (yield this.api.post(this.buildAuthGroupUri(), body))[1]; }));
        });
    }
    /**
     * Returns system information about the Geometry Minions.
     */
    getMinionInfo() {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, utils_1.sendRequest)(() => __awaiter(this, void 0, void 0, function* () { return (yield this.api.get(`${this.buildSystemUri()}/minions/info`))[1]; }));
        });
    }
}
exports.ShapeDiverSystemApi = ShapeDiverSystemApi;
