// @shapediver/viewer/dist/implementation/viewport/camera/OrthographicCameraApi.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _OrthographicCameraApi_camera, _OrthographicCameraApi_inputValidator, _OrthographicCameraApi_logger, _OrthographicCameraApi_viewportApi;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrthographicCameraApi = void 0;
const viewer_rendering_engine_camera_engine_1 = require("@shapediver/viewer.rendering-engine.camera-engine");
const AbstractCameraApi_1 = require("./AbstractCameraApi");
const viewer_shared_services_1 = require("@shapediver/viewer.shared.services");
const tsyringe_1 = require("tsyringe");
class OrthographicCameraApi extends AbstractCameraApi_1.AbstractCameraApi {
    // #endregion Properties (1)
    // #region Constructors (1)
    constructor(viewportApi, camera) {
        super(viewportApi, camera);
        // #region Properties (1)
        _OrthographicCameraApi_camera.set(this, void 0);
        _OrthographicCameraApi_inputValidator.set(this, tsyringe_1.container.resolve(viewer_shared_services_1.InputValidator));
        _OrthographicCameraApi_logger.set(this, tsyringe_1.container.resolve(viewer_shared_services_1.Logger));
        _OrthographicCameraApi_viewportApi.set(this, void 0);
        __classPrivateFieldSet(this, _OrthographicCameraApi_viewportApi, viewportApi, "f");
        __classPrivateFieldSet(this, _OrthographicCameraApi_camera, camera, "f");
        this.scope = 'OrthographicCameraApi';
    }
    // #endregion Constructors (1)
    // #region Public Accessors (18)
    get damping() {
        return __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.damping;
    }
    set damping(value) {
        const scope = 'damping';
        try {
            __classPrivateFieldGet(this, _OrthographicCameraApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.damping = value;
            __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _OrthographicCameraApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, e);
        }
    }
    get direction() {
        return __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").direction;
    }
    set direction(value) {
        const scope = 'direction';
        try {
            __classPrivateFieldGet(this, _OrthographicCameraApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, value, 'enum', true, Object.values(viewer_rendering_engine_camera_engine_1.ORTHOGRAPHIC_CAMERA_DIRECTION));
            __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").direction = value;
            __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _OrthographicCameraApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, e);
        }
    }
    get enableKeyPan() {
        return __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.enableKeyPan;
    }
    set enableKeyPan(value) {
        const scope = 'enableKeyPan';
        try {
            __classPrivateFieldGet(this, _OrthographicCameraApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, value, 'boolean');
            __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.enableKeyPan = value;
            __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _OrthographicCameraApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, e);
        }
    }
    get enablePan() {
        return __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.enablePan;
    }
    set enablePan(value) {
        const scope = 'enablePan';
        try {
            __classPrivateFieldGet(this, _OrthographicCameraApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, value, 'boolean');
            __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.enablePan = value;
            __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _OrthographicCameraApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, e);
        }
    }
    get enableZoom() {
        return __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.enableZoom;
    }
    set enableZoom(value) {
        const scope = 'enableZoom';
        try {
            __classPrivateFieldGet(this, _OrthographicCameraApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, value, 'boolean');
            __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.enableZoom = value;
            __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _OrthographicCameraApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, e);
        }
    }
    get keyPanSpeed() {
        return __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.keyPanSpeed;
    }
    set keyPanSpeed(value) {
        const scope = 'keyPanSpeed';
        try {
            __classPrivateFieldGet(this, _OrthographicCameraApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.keyPanSpeed = value;
            __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _OrthographicCameraApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, e);
        }
    }
    get movementSmoothness() {
        return __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.movementSmoothness;
    }
    set movementSmoothness(value) {
        const scope = 'movementSmoothness';
        try {
            __classPrivateFieldGet(this, _OrthographicCameraApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.movementSmoothness = value;
            __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _OrthographicCameraApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, e);
        }
    }
    get panSpeed() {
        return __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.panSpeed;
    }
    set panSpeed(value) {
        const scope = 'panSpeed';
        try {
            __classPrivateFieldGet(this, _OrthographicCameraApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.panSpeed = value;
            __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _OrthographicCameraApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, e);
        }
    }
    get zoomSpeed() {
        return __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.zoomSpeed;
    }
    set zoomSpeed(value) {
        const scope = 'zoomSpeed';
        try {
            __classPrivateFieldGet(this, _OrthographicCameraApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _OrthographicCameraApi_camera, "f").controls.zoomSpeed = value;
            __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _OrthographicCameraApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _OrthographicCameraApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.CAMERA, `${this.scope}.${scope}`, e);
        }
    }
}
exports.OrthographicCameraApi = OrthographicCameraApi;
_OrthographicCameraApi_camera = new WeakMap(), _OrthographicCameraApi_inputValidator = new WeakMap(), _OrthographicCameraApi_logger = new WeakMap(), _OrthographicCameraApi_viewportApi = new WeakMap();
