// @shapediver/api.geometry-api-dto-v2/dist/request/SdRequestDtoModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverRequestModelTrustlevel = exports.ShapeDiverRequestModelFiletype = void 0;
/**
 * @swagger {components.schemas} RequestModelFiletype {
     "description": "Supported file types for models.",
     "type": "string",
     "enum": [ "gh", "ghx" ]
    }
 */
/** Supported file types for models. */
var ShapeDiverRequestModelFiletype;
(function (ShapeDiverRequestModelFiletype) {
    ShapeDiverRequestModelFiletype["GRASSHOPPER_BINARY"] = "gh";
    ShapeDiverRequestModelFiletype["GRASSHOPPER_XML"] = "ghx";
})(ShapeDiverRequestModelFiletype = exports.ShapeDiverRequestModelFiletype || (exports.ShapeDiverRequestModelFiletype = {}));
/**
 * @swagger {components.schemas} RequestModelTrustlevel {
     "description": "Level of trust granted for a model.",
     "type": "string",
     "enum": [ "''", "none", "full" ]
    }
 */
/** Level of trust granted for a model. */
var ShapeDiverRequestModelTrustlevel;
(function (ShapeDiverRequestModelTrustlevel) {
    ShapeDiverRequestModelTrustlevel["UNDEFINED"] = "";
    ShapeDiverRequestModelTrustlevel["NONE"] = "none";
    ShapeDiverRequestModelTrustlevel["FULL"] = "full";
})(ShapeDiverRequestModelTrustlevel = exports.ShapeDiverRequestModelTrustlevel || (exports.ShapeDiverRequestModelTrustlevel = {}));
