// @shapediver/sdk.geometry-api-sdk-core/dist/resources/BaseResourceApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseResourceApi = void 0;
class BaseResourceApi {
    constructor(api) {
        this.commonPath = "/api/v2";
        this.api = api;
    }
    /**
     * Create the URI for analytics-based controller endpoints.
     *
     * @protected
     */
    buildAnalyticsUri() {
        return `${this.commonPath}/analytics`;
    }
    /**
     * Create the URI for AR-scene-based controller endpoints.
     *
     * @protected
     */
    buildArSceneUri() {
        return `${this.commonPath}/ar-scene`;
    }
    /**
     * Create the URI for auth_group-based controller endpoints.
     *
     * @protected
     */
    buildAuthGroupUri() {
        return `${this.commonPath}/auth_group`;
    }
    /**
     * Create the URI for ticket-based controller endpoints.
     *
     * @param ticketId
     * @protected
     */
    buildTicketUri(ticketId) {
        return `${this.commonPath}/ticket/${ticketId}`;
    }
    /**
     * Create the URI for session-based controller endpoints.
     *
     * @param sessionId
     * @protected
     */
    buildSessionUri(sessionId) {
        return `${this.commonPath}/session/${sessionId}`;
    }
    /**
     * Create the URI for model-based controller endpoints.
     *
     * @param modelId
     * @protected
     */
    buildModelUri(modelId) {
        return `${this.commonPath}/model/${modelId}`;
    }
    /**
     * Create the URI for system-based controller endpoints.
     *
     * @protected
     */
    buildSystemUri() {
        return `${this.commonPath}/system`;
    }
}
exports.BaseResourceApi = BaseResourceApi;
