// @shapediver/sdk.geometry-api-sdk-core/dist/api/ShapeDiverSdkApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverSdkApi = exports.ShapeDiverSdkApiResponseType = void 0;
const axios_1 = __importDefault(require("axios"));
const ShapeDiverErrors_1 = require("../ShapeDiverErrors");
var Method;
(function (Method) {
    Method["DELETE"] = "DELETE";
    Method["GET"] = "GET";
    Method["PATCH"] = "PATCH";
    Method["POST"] = "POST";
    Method["PUT"] = "PUT";
})(Method || (Method = {}));
var ShapeDiverSdkApiResponseType;
(function (ShapeDiverSdkApiResponseType) {
    ShapeDiverSdkApiResponseType["JSON"] = "json";
    ShapeDiverSdkApiResponseType["DATA"] = "arraybuffer";
    ShapeDiverSdkApiResponseType["TEXT"] = "text";
})(ShapeDiverSdkApiResponseType = exports.ShapeDiverSdkApiResponseType || (exports.ShapeDiverSdkApiResponseType = {}));
class ShapeDiverSdkApi {
    constructor(config) {
        this.config = config;
    }
    buildRequestConfig(method, options, data) {
        const request = {
            method: method,
            headers: {},
            responseType: options.responseType,
            data: undefined,
        };
        // Process HTTP headers
        if (!options.disableCustomHeaders)
            request.headers = Object.assign({}, this.config.headers);
        // Process HTTP authorization header
        if (options.disableAuthorization) {
            delete request.headers["Authorization"];
            delete request.headers["authorization"]; // config.headers might use lower case
        }
        else if (this.config.jwt) {
            request.headers["Authorization"] = "Bearer " + this.config.jwt;
        }
        // Set data and convert depending on content-type
        if (options.contentType) {
            request.headers["Content-Type"] = options.contentType;
            request.data = data;
        }
        // Set the HTTP Accept header if specified
        if (options.accept) {
            request.headers["Accept"] = options.accept;
        }
        return request;
    }
    buildUrl(uri) {
        if (typeof uri !== "string") {
            throw new ShapeDiverErrors_1.ShapeDiverError("No URL or URI was specified");
        }
        else if (uri.startsWith("http")) {
            return uri;
        }
        else if (uri.startsWith("/")) {
            return `${this.config.baseUrl}/${uri.substring(1)}`;
        }
        else {
            return `${this.config.baseUrl}/${uri}`;
        }
    }
    /**
     * Processes the given Axios error, maps it content and throws the
     * respective ShapeDiver error.
     *
     * @param error
     * @param responseType
     * @throws
     * @private
     */
    static processError(error, responseType) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function* () {
            if (error.response) {
                // Request was made and server responded with 4xx or 5xx
                const resp = error.response;
                let data;
                if (responseType === ShapeDiverSdkApiResponseType.DATA) {
                    data = this.convertErrorResponseData(resp.data);
                }
                else {
                    data = resp.data;
                }
                throw new ShapeDiverErrors_1.ShapeDiverResponseError(data.message || (data.desc || ((_a = data.error) !== null && _a !== void 0 ? _a : "")), resp.status, (_b = data.error) !== null && _b !== void 0 ? _b : "", (_c = data.desc) !== null && _c !== void 0 ? _c : "", resp.headers);
            }
            else if (error.request) {
                // The request was made but no response was received
                throw new ShapeDiverErrors_1.ShapeDiverRequestError("Could not send request.", "The request was made but no response was received");
            }
            else {
                // Something happened in setting up the request that triggered an Error
                throw new ShapeDiverErrors_1.ShapeDiverError(error.message);
            }
        });
    }
    /**
     * Axios returns errors in the same type that we where using to specify the
     * data response type of the happy-path. Thus, we have to convert them
     * manually.
     *
     * @param data
     * @private
     */
    static convertErrorResponseData(data) {
        let stringData;
        if (typeof window === "undefined") {
            // NodeJs
            stringData = Buffer.from(data).toString();
        }
        else if (window.TextDecoder) {
            // Browser + TextDecoder support
            stringData = new TextDecoder("utf-8").decode(new Uint8Array(data));
        }
        else {
            throw new ShapeDiverErrors_1.ShapeDiverError("Received an unknown error object");
        }
        try {
            // This should work for the ShapeDiver backend
            return JSON.parse(stringData);
        }
        catch (_) {
            // This might be an XML when calling another system (e.g. S3)
            return { message: stringData };
        }
    }
    get(url, options = {
        contentType: "application/json",
        responseType: ShapeDiverSdkApiResponseType.JSON,
    }) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = this.buildRequestConfig(Method.GET, options, undefined);
            try {
                const response = yield (0, axios_1.default)(this.buildUrl(url), config);
                return [response.headers, response.data];
            }
            catch (e) {
                return yield ShapeDiverSdkApi.processError(e, options.responseType);
            }
        });
    }
    post(url, data = {}, options = {
        contentType: "application/json",
        responseType: ShapeDiverSdkApiResponseType.JSON,
    }) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = this.buildRequestConfig(Method.POST, options, data);
            try {
                const response = yield (0, axios_1.default)(this.buildUrl(url), config);
                return [response.headers, response.data];
            }
            catch (e) {
                return yield ShapeDiverSdkApi.processError(e, options.responseType);
            }
        });
    }
    put(url, data = {}, options = {
        contentType: "application/json",
        responseType: ShapeDiverSdkApiResponseType.JSON,
    }) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = this.buildRequestConfig(Method.PUT, options, data);
            try {
                const response = yield (0, axios_1.default)(this.buildUrl(url), config);
                return [response.headers, response.data];
            }
            catch (e) {
                return yield ShapeDiverSdkApi.processError(e, options.responseType);
            }
        });
    }
    patch(url, data = {}, options = {
        contentType: "application/json",
        responseType: ShapeDiverSdkApiResponseType.JSON,
    }) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = this.buildRequestConfig(Method.PATCH, options, data);
            try {
                const response = yield (0, axios_1.default)(this.buildUrl(url), config);
                return [response.headers, response.data];
            }
            catch (e) {
                return yield ShapeDiverSdkApi.processError(e, options.responseType);
            }
        });
    }
    delete(url, options = {
        contentType: "application/json",
        responseType: ShapeDiverSdkApiResponseType.JSON,
    }) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = this.buildRequestConfig(Method.DELETE, options, {});
            try {
                const response = yield (0, axios_1.default)(this.buildUrl(url), config);
                return [response.headers, response.data];
            }
            catch (e) {
                return yield ShapeDiverSdkApi.processError(e, options.responseType);
            }
        });
    }
}
exports.ShapeDiverSdkApi = ShapeDiverSdkApi;
