// @shapediver/viewer.shared.types/src/index.ts
import { CustomData } from './implementation/data/CustomData'
import { AttributeData, GeometryData, PrimitiveData } from './implementation/data/GeometryData'
import {
  MaterialStandardData,
} from './implementation/material/MaterialStandardData'
import { HTMLElementAnchorCustomData, HTMLElementAnchorData, HTMLElementAnchorImageData, HTMLElementAnchorTextData } from './implementation/data/HTMLElementAnchorData'
import { SDTFItemData } from './implementation/sdtf/SDTFItemData'
import { AnimationData } from './implementation/data/AnimationData'
import { IViewportEvent } from './interfaces/events/IViewportEvent'
import { ISessionEvent } from './interfaces/events/ISessionEvent'
import { ICameraEvent } from './interfaces/events/ICameraEvent'
import { ISceneEvent } from './interfaces/events/ISceneEvent'
import { ITaskEvent, TASK_TYPE } from './interfaces/events/ITaskEvent'
import { MaterialVariantsData } from './implementation/material/MaterialVariantsData'
import { MapData } from './implementation/material/MapData'
import { MaterialSpecularGlossinessData } from './implementation/material/MaterialSpecularGlossinessData'
import { MaterialUnlitData } from './implementation/material/MaterialUnlitData'
import { IAnchorDataImage, IAnchorDataText, IHTMLElementAnchorData } from './interfaces/data/IHTMLElementAnchorData'
import { IMaterialAbstractData, IMaterialAbstractDataProperties, MATERIAL_ALPHA, MATERIAL_SHADING, MATERIAL_SIDE } from './interfaces/data/material/IMaterialAbstractData'
import { IAttributeData, IGeometryData, IPrimitiveData, PRIMITIVE_MODE } from './interfaces/data/IGeometryData'
import { TEXTURE_WRAPPING, TEXTURE_FILTERING, IMapData } from './interfaces/data/material/IMapData'
import { IMaterialSpecularGlossinessData, IMaterialSpecularGlossinessDataProperties } from './interfaces/data/material/IMaterialSpecularGlossinessDataProperties'
import { IMaterialStandardData, IMaterialStandardDataProperties } from './interfaces/data/material/IMaterialStandardData'
import { IMaterialUnlitData, IMaterialUnlitDataProperties } from './interfaces/data/material/IMaterialUnlitData'
import { IMaterialVariantsData } from './interfaces/data/material/IMaterialVariantsData'
import { IAnimationData, IAnimationTrack } from './interfaces/data/IAnimationData'
import { ICustomData } from './interfaces/data/ICustomData'
import { AbstractMaterialData } from './implementation/material/AbstractMaterialData'
import { ISDTFOverview, ISDTFOverviewData } from './interfaces/sdtf/ISDTFOverviewData'
import { SDTFAttributeData, SDTFAttributesData } from './implementation/sdtf/SDTFAttributesData'
import { SDTFOverviewData } from './implementation/sdtf/SDTFOverviewData'
import { ISDTFAttributesData, ISDTFAttributeData } from './interfaces/sdtf/ISDTFAttributesData'
import { ISDTFItemData } from './interfaces/sdtf/ISDTFItemData'
import { ISDTFAttributeVisualizationData } from './interfaces/sdtf/ISDTFAttributeVisualizationData'
import { EventResponseMapping } from './interfaces/events/EventResponseMapping'
import { IBoneData } from './interfaces/data/IBoneData'
import { BoneData } from './implementation/data/BoneData'
import { SdtfTypeHintName } from '@shapediver/sdk.sdtf-v1'
import { SdtfPrimitiveTypeGuard } from '@shapediver/sdk.sdtf-primitives'
import { MaterialGemData } from './implementation/material/MaterialGemData'
import { IMaterialGemData, IMaterialGemDataProperties } from './interfaces/data/material/IMaterialGemDataProperties'
import { MaterialShadowData } from './implementation/material/MaterialShadowData'
import { IMaterialShadowData, IMaterialShadowDataProperties } from './interfaces/data/material/IMaterialShadowData'

export {
  IMaterialStandardData, MaterialStandardData, IMaterialStandardDataProperties, 
  IMaterialAbstractData, IMaterialAbstractDataProperties, AbstractMaterialData,
  IMaterialUnlitData, MaterialUnlitData, IMaterialUnlitDataProperties, 
  IMaterialShadowData, MaterialShadowData, IMaterialShadowDataProperties, 
  IMaterialSpecularGlossinessData, MaterialSpecularGlossinessData, IMaterialSpecularGlossinessDataProperties, 
  IMaterialGemData, MaterialGemData, IMaterialGemDataProperties, 
  IMapData, MapData, MATERIAL_SIDE, MATERIAL_ALPHA, MATERIAL_SHADING, TEXTURE_WRAPPING, TEXTURE_FILTERING
}

export {
  IAnimationData, AnimationData, IAnimationTrack, IGeometryData, IAttributeData, IPrimitiveData, IMaterialVariantsData, GeometryData, AttributeData, PrimitiveData, MaterialVariantsData, PRIMITIVE_MODE
}

export {
  IAnchorDataImage, IAnchorDataText, IHTMLElementAnchorData, HTMLElementAnchorCustomData, HTMLElementAnchorTextData, HTMLElementAnchorImageData, HTMLElementAnchorData
}

export {
  ICustomData, CustomData, IBoneData, BoneData
}

export {
  EventResponseMapping, IViewportEvent, ISessionEvent, ICameraEvent, ISceneEvent, ITaskEvent, TASK_TYPE
}

export {
  ISDTFOverviewData, SDTFOverviewData, ISDTFOverview, SDTFAttributesData, ISDTFAttributesData, ISDTFAttributeData, SDTFAttributeData, SDTFItemData, ISDTFItemData, ISDTFAttributeVisualizationData,
  SdtfTypeHintName as SDTF_TYPEHINT, SdtfPrimitiveTypeGuard
}
