// @shapediver/viewer.session-engine.session-engine/dist/interfaces/ISessionEngine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PARAMETER_VISUALIZATION = exports.PARAMETER_TYPE = void 0;
/**
 * The type of the parameter.
 */
var PARAMETER_TYPE;
(function (PARAMETER_TYPE) {
    PARAMETER_TYPE["FLOAT"] = "Float";
    PARAMETER_TYPE["INT"] = "Int";
    PARAMETER_TYPE["EVEN"] = "Even";
    PARAMETER_TYPE["ODD"] = "Odd";
    PARAMETER_TYPE["STRING"] = "String";
    PARAMETER_TYPE["COLOR"] = "Color";
    PARAMETER_TYPE["STRINGLIST"] = "StringList";
    PARAMETER_TYPE["BOOL"] = "Bool";
    PARAMETER_TYPE["TIME"] = "Time";
    PARAMETER_TYPE["FILE"] = "File";
    PARAMETER_TYPE["SNUMBER"] = "sNumber";
    PARAMETER_TYPE["SINTEGER"] = "sInteger";
    PARAMETER_TYPE["SSTRING"] = "sString";
    PARAMETER_TYPE["SCOLOR"] = "sColor";
    PARAMETER_TYPE["SBOOL"] = "sBool";
    PARAMETER_TYPE["STIME"] = "sTime";
    PARAMETER_TYPE["SBITMAP"] = "sBitmap";
    PARAMETER_TYPE["SDOMAIN"] = "sDomain";
    PARAMETER_TYPE["SDOMAIN2D"] = "sDomain2D";
    PARAMETER_TYPE["SPOINT"] = "sPoint";
    PARAMETER_TYPE["SLINE"] = "sLine";
    PARAMETER_TYPE["SVECTOR"] = "sVector";
    PARAMETER_TYPE["SBOX"] = "sBox";
    PARAMETER_TYPE["SPLANE"] = "sPlane";
    PARAMETER_TYPE["SRECTANGLE"] = "sRectangle";
    PARAMETER_TYPE["SCURVE"] = "sCurve";
    PARAMETER_TYPE["SCIRCLE"] = "sCircle";
    PARAMETER_TYPE["SMESH"] = "sMesh";
    PARAMETER_TYPE["SSURFACE"] = "sSurface";
    PARAMETER_TYPE["SBREP"] = "sBrep";
    PARAMETER_TYPE["SSUBDIV"] = "sSubdiv";
})(PARAMETER_TYPE = exports.PARAMETER_TYPE || (exports.PARAMETER_TYPE = {}));
/**
 * Type of visualization which should be used for UI elements representing the parameter.
 */
var PARAMETER_VISUALIZATION;
(function (PARAMETER_VISUALIZATION) {
    PARAMETER_VISUALIZATION["SLIDER"] = "slider";
    PARAMETER_VISUALIZATION["SEQUENCE"] = "sequence";
    PARAMETER_VISUALIZATION["CYCLE"] = "cycle";
    PARAMETER_VISUALIZATION["DROPDOWN"] = "dropdown";
    PARAMETER_VISUALIZATION["CHECKLIST"] = "checklist";
    PARAMETER_VISUALIZATION["CLOCK"] = "color";
    PARAMETER_VISUALIZATION["CALENDAR"] = "calendar";
    PARAMETER_VISUALIZATION["TOGGLE"] = "toggle";
    PARAMETER_VISUALIZATION["SWATCH"] = "swatch";
    PARAMETER_VISUALIZATION["BUTTON"] = "button";
    PARAMETER_VISUALIZATION["DIAL"] = "dial";
    PARAMETER_VISUALIZATION["TEXT"] = "text";
})(PARAMETER_VISUALIZATION = exports.PARAMETER_VISUALIZATION || (exports.PARAMETER_VISUALIZATION = {}));
;
