// @shapediver/viewer/dist/implementation/viewport/lights/types/DirectionalLightApi.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _DirectionalLightApi_inputValidator, _DirectionalLightApi_light, _DirectionalLightApi_logger, _DirectionalLightApi_viewportApi;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectionalLightApi = void 0;
const AbstractLightApi_1 = require("../AbstractLightApi");
const viewer_shared_services_1 = require("@shapediver/viewer.shared.services");
const tsyringe_1 = require("tsyringe");
class DirectionalLightApi extends AbstractLightApi_1.AbstractLightApi {
    // #endregion Properties (4)
    // #region Constructors (1)
    constructor(viewportApi, light) {
        super(viewportApi, light);
        // #region Properties (4)
        _DirectionalLightApi_inputValidator.set(this, tsyringe_1.container.resolve(viewer_shared_services_1.InputValidator));
        _DirectionalLightApi_light.set(this, void 0);
        _DirectionalLightApi_logger.set(this, tsyringe_1.container.resolve(viewer_shared_services_1.Logger));
        _DirectionalLightApi_viewportApi.set(this, void 0);
        __classPrivateFieldSet(this, _DirectionalLightApi_viewportApi, viewportApi, "f");
        __classPrivateFieldSet(this, _DirectionalLightApi_light, light, "f");
        this.scope = 'DirectionalLightApi';
    }
    // #endregion Constructors (1)
    // #region Public Accessors (8)
    get castShadow() {
        return __classPrivateFieldGet(this, _DirectionalLightApi_light, "f").castShadow;
    }
    set castShadow(value) {
        const scope = 'castShadow';
        try {
            __classPrivateFieldGet(this, _DirectionalLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'boolean');
            __classPrivateFieldGet(this, _DirectionalLightApi_light, "f").castShadow = value;
            __classPrivateFieldGet(this, _DirectionalLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _DirectionalLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _DirectionalLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
    get direction() {
        return __classPrivateFieldGet(this, _DirectionalLightApi_light, "f").direction;
    }
    set direction(value) {
        const scope = 'direction';
        try {
            __classPrivateFieldGet(this, _DirectionalLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'vec3');
            __classPrivateFieldGet(this, _DirectionalLightApi_light, "f").direction = value;
            __classPrivateFieldGet(this, _DirectionalLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _DirectionalLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _DirectionalLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
    get shadowMapBias() {
        return __classPrivateFieldGet(this, _DirectionalLightApi_light, "f").shadowMapBias;
    }
    set shadowMapBias(value) {
        const scope = 'shadowMapBias';
        try {
            __classPrivateFieldGet(this, _DirectionalLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _DirectionalLightApi_light, "f").shadowMapBias = value;
            __classPrivateFieldGet(this, _DirectionalLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _DirectionalLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _DirectionalLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
    get shadowMapResolution() {
        return __classPrivateFieldGet(this, _DirectionalLightApi_light, "f").shadowMapResolution;
    }
    set shadowMapResolution(value) {
        const scope = 'shadowMapResolution';
        try {
            __classPrivateFieldGet(this, _DirectionalLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _DirectionalLightApi_light, "f").shadowMapResolution = value;
            __classPrivateFieldGet(this, _DirectionalLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _DirectionalLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _DirectionalLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
}
exports.DirectionalLightApi = DirectionalLightApi;
_DirectionalLightApi_inputValidator = new WeakMap(), _DirectionalLightApi_light = new WeakMap(), _DirectionalLightApi_logger = new WeakMap(), _DirectionalLightApi_viewportApi = new WeakMap();
