// @shapediver/viewer/dist/implementation/viewport/lights/types/SpotLightApi.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SpotLightApi_light, _SpotLightApi_inputValidator, _SpotLightApi_logger, _SpotLightApi_viewportApi;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpotLightApi = void 0;
const AbstractLightApi_1 = require("../AbstractLightApi");
const viewer_shared_services_1 = require("@shapediver/viewer.shared.services");
const tsyringe_1 = require("tsyringe");
class SpotLightApi extends AbstractLightApi_1.AbstractLightApi {
    // #endregion Properties (7)
    // #region Constructors (1)
    constructor(viewportApi, light) {
        super(viewportApi, light);
        // #region Properties (7)
        _SpotLightApi_light.set(this, void 0);
        _SpotLightApi_inputValidator.set(this, tsyringe_1.container.resolve(viewer_shared_services_1.InputValidator));
        _SpotLightApi_logger.set(this, tsyringe_1.container.resolve(viewer_shared_services_1.Logger));
        _SpotLightApi_viewportApi.set(this, void 0);
        __classPrivateFieldSet(this, _SpotLightApi_viewportApi, viewportApi, "f");
        __classPrivateFieldSet(this, _SpotLightApi_light, light, "f");
        this.scope = 'SpotLightApi';
    }
    // #endregion Constructors (1)
    // #region Public Accessors (12)
    get angle() {
        return __classPrivateFieldGet(this, _SpotLightApi_light, "f").angle;
    }
    set angle(value) {
        const scope = 'angle';
        try {
            __classPrivateFieldGet(this, _SpotLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _SpotLightApi_light, "f").angle = value;
            __classPrivateFieldGet(this, _SpotLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _SpotLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _SpotLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
    get decay() {
        return __classPrivateFieldGet(this, _SpotLightApi_light, "f").decay;
    }
    set decay(value) {
        const scope = 'decay';
        try {
            __classPrivateFieldGet(this, _SpotLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _SpotLightApi_light, "f").decay = value;
            __classPrivateFieldGet(this, _SpotLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _SpotLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _SpotLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
    get distance() {
        return __classPrivateFieldGet(this, _SpotLightApi_light, "f").distance;
    }
    set distance(value) {
        const scope = 'distance';
        try {
            __classPrivateFieldGet(this, _SpotLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _SpotLightApi_light, "f").distance = value;
            __classPrivateFieldGet(this, _SpotLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _SpotLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _SpotLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
    get penumbra() {
        return __classPrivateFieldGet(this, _SpotLightApi_light, "f").penumbra;
    }
    set penumbra(value) {
        const scope = 'penumbra';
        try {
            __classPrivateFieldGet(this, _SpotLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'number');
            __classPrivateFieldGet(this, _SpotLightApi_light, "f").penumbra = value;
            __classPrivateFieldGet(this, _SpotLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _SpotLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _SpotLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
    get position() {
        return __classPrivateFieldGet(this, _SpotLightApi_light, "f").position;
    }
    set position(value) {
        const scope = 'position';
        try {
            __classPrivateFieldGet(this, _SpotLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'vec3');
            __classPrivateFieldGet(this, _SpotLightApi_light, "f").position = value;
            __classPrivateFieldGet(this, _SpotLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _SpotLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _SpotLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
    get target() {
        return __classPrivateFieldGet(this, _SpotLightApi_light, "f").target;
    }
    set target(value) {
        const scope = 'target';
        try {
            __classPrivateFieldGet(this, _SpotLightApi_inputValidator, "f").validateAndError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, value, 'vec3');
            __classPrivateFieldGet(this, _SpotLightApi_light, "f").target = value;
            __classPrivateFieldGet(this, _SpotLightApi_logger, "f").debug(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}: ${scope} was set to: ${value}`);
            __classPrivateFieldGet(this, _SpotLightApi_viewportApi, "f").update();
        }
        catch (e) {
            if (e instanceof viewer_shared_services_1.ShapeDiverViewerError || e instanceof viewer_shared_services_1.ShapeDiverBackendError)
                throw e;
            throw __classPrivateFieldGet(this, _SpotLightApi_logger, "f").handleError(viewer_shared_services_1.LOGGING_TOPIC.LIGHT, `${this.scope}.${scope}`, e);
        }
    }
}
exports.SpotLightApi = SpotLightApi;
_SpotLightApi_light = new WeakMap(), _SpotLightApi_inputValidator = new WeakMap(), _SpotLightApi_logger = new WeakMap(), _SpotLightApi_viewportApi = new WeakMap();
