// @shapediver/sdk.geometry-api-sdk-v2/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverSdk = exports.create = exports.ShapeDiverResponseError = exports.ShapeDiverSdkConfigType = exports.ShapeDiverSdkApiResponseType = exports.ShapeDiverRequestError = exports.ShapeDiverError = void 0;
__exportStar(require("@shapediver/api.geometry-api-dto-v2"), exports);
var sdk_geometry_api_sdk_core_1 = require("@shapediver/sdk.geometry-api-sdk-core");
Object.defineProperty(exports, "ShapeDiverError", { enumerable: true, get: function () { return sdk_geometry_api_sdk_core_1.ShapeDiverError; } });
Object.defineProperty(exports, "ShapeDiverRequestError", { enumerable: true, get: function () { return sdk_geometry_api_sdk_core_1.ShapeDiverRequestError; } });
Object.defineProperty(exports, "ShapeDiverSdkApiResponseType", { enumerable: true, get: function () { return sdk_geometry_api_sdk_core_1.ShapeDiverSdkApiResponseType; } });
Object.defineProperty(exports, "ShapeDiverSdkConfigType", { enumerable: true, get: function () { return sdk_geometry_api_sdk_core_1.ShapeDiverSdkConfigType; } });
var ShapeDiverErrors_1 = require("./ShapeDiverErrors");
Object.defineProperty(exports, "ShapeDiverResponseError", { enumerable: true, get: function () { return ShapeDiverErrors_1.ShapeDiverResponseError; } });
var ShapeDiverSdk_1 = require("./ShapeDiverSdk");
Object.defineProperty(exports, "create", { enumerable: true, get: function () { return ShapeDiverSdk_1.create; } });
Object.defineProperty(exports, "ShapeDiverSdk", { enumerable: true, get: function () { return ShapeDiverSdk_1.ShapeDiverSdk; } });
