// @shapediver/api.geometry-api-dto-v2/dist/response/SdResponseDtoModelComputation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverResponseModelComputationStatus = void 0;
/**
* @swagger {components.schemas} ResponseModelComputationStatus {
    "description": "Result of a computation request.",
        "type": "string",
        "enum": [ "success", "timeout", "checkconfirmed", "checkdenied", "checkpending", "maxcombinedassetsizeexceeded", "maxdbsizeperoutputexceeded", "maxpartsperoutputexceeded", "maxassetpartsperoutputexceeded", "maxtransformationsperoutputexceeded", "maxpartsexceeded", "maxassetpartsexceeded", "recoverableerror", "unrecoverableerror", "nooutputdatafordefaultparametervalues", "modelwithoutgeometryoutput", "unknown" ]
}
*/
/** Result of a computation request.*/
var ShapeDiverResponseModelComputationStatus;
(function (ShapeDiverResponseModelComputationStatus) {
    /** Model computation was successful */
    ShapeDiverResponseModelComputationStatus["SUCCESS"] = "success";
    /** Model computation timed out */
    ShapeDiverResponseModelComputationStatus["TIMEOUT"] = "timeout";
    /** Model was confirmed */
    ShapeDiverResponseModelComputationStatus["CHECK_CONFIRMED"] = "checkconfirmed";
    /** Model was denied */
    ShapeDiverResponseModelComputationStatus["CHECK_DENIED"] = "checkdenied";
    /** Model check is pending (e.g. due to manual script checking required) */
    ShapeDiverResponseModelComputationStatus["CHECK_PENDING"] = "checkpending";
    /** Maximum size for data to be written to files exceeded */
    ShapeDiverResponseModelComputationStatus["MAX_COMBINED_ASSET_SIZE_EXCEEDED"] = "maxcombinedassetsizeexceeded";
    /** Maximum output size for data to be written to the database exceeded */
    ShapeDiverResponseModelComputationStatus["MAX_DB_SIZE_PER_OUTPUT_EXCEEDED"] = "maxdbsizeperoutputexceeded";
    /** Maximum number of allowed data parts per output exceeded */
    ShapeDiverResponseModelComputationStatus["MAX_PARTS_PER_OUTPUT_EXCEEDED"] = "maxpartsperoutputexceeded";
    /** Maximum number of allowed file parts per output exceeded */
    ShapeDiverResponseModelComputationStatus["MAX_ASSET_PARTS_PER_OUTPUT_EXCEEDED"] = "maxassetpartsperoutputexceeded";
    /** Maximum number of allowed transformations per output exceeded */
    ShapeDiverResponseModelComputationStatus["MAX_TRANSFORMATIONS_PER_OUTPUT_EXCEEDED"] = "maxtransformationsperoutputexceeded";
    /** Maximum number of allowed data parts per work order exceeded */
    ShapeDiverResponseModelComputationStatus["MAX_PARTS_EXCEEDED"] = "maxpartsexceeded";
    /** Maximum number of allowed file parts per work order exceeded */
    ShapeDiverResponseModelComputationStatus["MAX_ASSET_PARTS_EXCEEDED"] = "maxassetpartsexceeded";
    /** Recoverable error while processing the request */
    ShapeDiverResponseModelComputationStatus["RECOVERABLE_ERROR"] = "recoverableerror";
    /** Unrecoverable error while processing the request */
    ShapeDiverResponseModelComputationStatus["UNRECOVERABLE_ERROR"] = "unrecoverableerror";
    /** The model does not have output data for its default parameter values, and the worker is configured to not allow this. */
    ShapeDiverResponseModelComputationStatus["NO_OUTPUT_DATA_FOR_DEFAULT_PARAMETER_VALUES"] = "nooutputdatafordefaultparametervalues";
    /** The model does not contain any outputs which could output geometry, and the worker is configured to not allow this. */
    ShapeDiverResponseModelComputationStatus["MODEL_WITHOUT_GEOMETRY_OUTPUT"] = "modelwithoutgeometryoutput";
    /** Unknown status */
    ShapeDiverResponseModelComputationStatus["UNKNOWN"] = "unknown";
})(ShapeDiverResponseModelComputationStatus = exports.ShapeDiverResponseModelComputationStatus || (exports.ShapeDiverResponseModelComputationStatus = {}));
