// @shapediver/api.geometry-api-dto-v2/dist/response/SdResponseDtoMisc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShapeDiverResponseModelCleanupProcessType = exports.ShapeDiverResponseTicketType = exports.ShapeDiverResponseParameterStructure = void 0;
const SdCommonsTicket_1 = require("../commons/SdCommonsTicket");
/**
 * @swagger {components.schemas} ResponseParameterStructure {
     "description": "Structural information of a parameter.",
     "type": "string",
     "enum": [ "item", "list", "tree" ]
   }
 */
/** Structural information of a parameter. */
var ShapeDiverResponseParameterStructure;
(function (ShapeDiverResponseParameterStructure) {
    ShapeDiverResponseParameterStructure["ITEM"] = "item";
    ShapeDiverResponseParameterStructure["LIST"] = "list";
    ShapeDiverResponseParameterStructure["TREE"] = "tree";
})(ShapeDiverResponseParameterStructure = exports.ShapeDiverResponseParameterStructure || (exports.ShapeDiverResponseParameterStructure = {}));
/**
 * @swagger {components.schemas} ResponseTicketType {
     "allOf": [
       { "$ref": "#/components/schemas/CommonsTicketType" }
     ]
   }
 */
/** */
exports.ShapeDiverResponseTicketType = Object.assign({}, SdCommonsTicket_1.ShapeDiverCommonsTicketType);
/**
 * @swagger {components.schemas} ResponseModelCleanupProcessType {
     "description": "Model cleanup process types.",
     "type": "string",
     "enum": [ "delete_export_version", "delete_model_texture", "delete_output_version" ]
   }
 */
/** Model cleanup process types. */
var ShapeDiverResponseModelCleanupProcessType;
(function (ShapeDiverResponseModelCleanupProcessType) {
    ShapeDiverResponseModelCleanupProcessType["DELETE_EXPORT_VERSION"] = "delete_export_version";
    ShapeDiverResponseModelCleanupProcessType["DELETE_MODEL_TEXTURE"] = "delete_model_texture";
    ShapeDiverResponseModelCleanupProcessType["DELETE_OUTPUT_VERSION"] = "delete_output_version";
})(ShapeDiverResponseModelCleanupProcessType = exports.ShapeDiverResponseModelCleanupProcessType || (exports.ShapeDiverResponseModelCleanupProcessType = {}));
